client/beefy/src/tests.rs

```diff
@@ -0,0 +1,442 @@
+	keystore
+}
+
+fn create_grandpa_keystore(authority: Ed25519Keyring) -> (SyncCryptoStorePtr, tempfile::TempDir) {
+	let keystore_path = tempfile::tempdir().expect("Creates keystore path");
+	let keystore =
+		Arc::new(LocalKeystore::open(keystore_path.path(), None).expect("Creates keystore"));
+	SyncCryptoStore::ed25519_generate_new(&*keystore, GRANDPA, Some(&authority.to_seed()))
+		.expect("Creates authority key");
+
+	(keystore, keystore_path)
+}
+
+// Spawns grandpa voters. Returns a future to spawn on the runtime.
+fn initialize_grandpa(
+	net: &mut BeefyTestNet,
+	peers: &[Ed25519Keyring],
+) -> impl Future<Output = ()> {
+	let voters = FuturesUnordered::new();
+
+	for (peer_id, key) in peers.iter().enumerate() {
+		let (keystore, _) = create_grandpa_keystore(*key);
+
+		let (net_service, link) = {
+			// temporary needed for some reason
+			let link = net.peers[peer_id]
+				.data
+				.grandpa_link_half
+				.lock()
+				.take()
+				.expect("link initialized at startup; qed");
+			(net.peers[peer_id].network_service().clone(), link)
+		};
+
+		let grandpa_params = GrandpaParams {
+			config: Config {
+				gossip_duration: TEST_GOSSIP_DURATION,
+				justification_period: 32,
+				keystore: Some(keystore),
+				name: Some(format!("peer#{}", peer_id)),
+				local_role: Role::Authority,
+				observer_enabled: true,
+				telemetry: None,
+				protocol_name: GRANDPA_PROTOCOL_NAME.into(),
+			},
+			link,
+			network: net_service,
+			voting_rule: (),
+			prometheus_registry: None,
+			shared_voter_state: SharedVoterState::empty(),
+			telemetry: None,
+		};
+		let voter =
+			run_grandpa_voter(grandpa_params).expect("all in order with client and network");
+
+		fn assert_send<T: Send>(_: &T) {}
+		assert_send(&voter);
+
+		voters.push(voter);
+	}
+
+	voters.for_each(|_| async move {})
+}
+
+// Spawns beefy voters. Returns a future to spawn on the runtime.
+fn initialize_beefy(net: &mut BeefyTestNet, peers: &[BeefyKeyring]) -> impl Future<Output = ()> {
+	let voters = FuturesUnordered::new();
+
+	for (peer_id, key) in peers.iter().enumerate() {
+		let keystore = create_beefy_keystore(*key);
+
+		let (signed_commitment_sender, signed_commitment_stream) =
+			BeefySignedCommitmentStream::<Block>::channel();
+		let (beefy_best_block_sender, beefy_best_block_stream) =
+			BeefyBestBlockStream::<Block>::channel();
+
+		let beefy_link_half = BeefyLinkHalf { signed_commitment_stream, beefy_best_block_stream };
+		*net.peers[peer_id].data.beefy_link_half.lock() = Some(beefy_link_half);
+
+		let beefy_params = crate::BeefyParams {
+			client: net.peers[peer_id].client().as_client(),
+			backend: net.peers[peer_id].client().as_backend(),
+			runtime: net.test_api.clone(),
+			key_store: Some(keystore),
+			network: net.peers[peer_id].network_service().clone(),
+			signed_commitment_sender,
+			beefy_best_block_sender,
+			min_block_delta: 4,
+			prometheus_registry: None,
+			protocol_name: BEEFY_PROTOCOL_NAME.into(),
```

Comment:
Votes don't seem to be gossiped for this `protocol_name`... How do I whitelist it?